homeassistant/helpers/event.py

```diff
@@ -103,12 +100,94 @@ def state_change_listener(event: Event) -> None:
                 event.data.get("new_state"),
             )
 
+    if entity_ids != MATCH_ALL:
+        # If we have a list of entity ids we use
+        # async_track_state_change_event to route
+        # by entity_id to avoid iterating though state change
+        # events and creating a jobs where the most
+        # common outcome is to return right away because
+        # the entity_id does not match since usually
+        # only one or two listeners want that specific
+        # entity_id.
+        return async_track_state_change_event(hass, entity_ids, state_change_listener)
+
     return hass.bus.async_listen(EVENT_STATE_CHANGED, state_change_listener)
 
 
 track_state_change = threaded_listener_factory(async_track_state_change)
 
 
+@bind_hass
+def async_track_state_change_event(
+    hass: HomeAssistant, entity_ids: Iterable[str], action: Callable[[Event], None]
+) -> Callable[[], None]:
+    """Track specific state change events indexed by entity_id.
+
+    Unlike async_track_state_change, async_track_state_change_event
+    passes the full event to the callback.
+
+    In order to avoid having to iterate a long list
+    of EVENT_STATE_CHANGED and fire and create a job
+    for each one, we keep a dict of entity ids that
+    care about the state change events so we can
+    do a fast dict lookup to route events.
+    """
+
+    if TRACK_STATE_CHANGE_CALLBACKS not in hass.data:
+        hass.data[TRACK_STATE_CHANGE_CALLBACKS] = {}
+
+    entity_callbacks = hass.data[TRACK_STATE_CHANGE_CALLBACKS]
```

Comment:
`entity_callbacks = hass.data[TRACK_STATE_CHANGE_CALLBACKS].setdefault({})` so you can save the line above

Reply:
done